listenbrainz/domain/spotify.py

```diff
@@ -0,0 +1,150 @@
+import pytz
+from flask import current_app, url_for
+import spotipy.oauth2
+
+from listenbrainz.db import spotify as db_spotify
+import datetime
+
+
+class Spotify:
+    def __init__(self, user_id, musicbrainz_id, user_token, token_expires, refresh_token,
+                 last_updated, active, update_error_message):
+        self.user_id = user_id
+        self.user_token = user_token
+        self.token_expires = token_expires
+        self.refresh_token = refresh_token
+        self.last_updated = last_updated
+        self.active = active
+        self.update_error_message = update_error_message
+        self.musicbrainz_id = musicbrainz_id
+
+    def get_spotipy_client(self):
+        return spotipy.Spotify(auth=self.user_token)
+
+    def nice_last_updated(self):
+        if not self.last_updated:
+            return 'never'
+        return self.last_updated
+
+    @property
+    def token_expired(self):
+        now = datetime.datetime.utcnow()
+        now = now.replace(tzinfo=pytz.UTC)
+        return now >= self.token_expires
+
+    @staticmethod
+    def from_dbrow(row):
+        return Spotify(
+           user_id=row['user_id'],
+           user_token=row['user_token'],
+           token_expires=row['token_expires'],
+           refresh_token=row['refresh_token'],
+           last_updated=row['last_updated'],
+           active=row['active'],
+           update_error_message=row['update_error'],
+           musicbrainz_id=row['musicbrainz_id'],
+        )
+
+    def __str__(self):
+        return "<Spotify(user:%s): %s>" % (self.user_id, self.musicbrainz_id)
+
+
+def refresh_user_token(spotify_user):
```

Comment:
Should there be any sort of validation on spotify_user?

Reply:
I don't think it's needed. Bad values would lead to raising `SpotifyException` which would be caught later in the script anyways.

Reply:
additionally, this is only called by us, with a value from the database, originally populated by an API call to spotify, so I don't think it's is actually possible to have a bad value here